subsys/bluetooth/host/avdtp.c

```diff
@@ -43,24 +46,883 @@ static struct bt_avdtp_seid_lsep *lseps;
 
 #define AVDTP_TIMEOUT K_SECONDS(6)
 
+K_MUTEX_DEFINE(avdtp_mutex);
+#define AVDTP_LOCK() k_mutex_lock(&avdtp_mutex, K_FOREVER)
+#define AVDTP_UNLOCK() k_mutex_unlock(&avdtp_mutex)
+
+static void avdtp_discover_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type,
+			uint8_t tid);
+static void avdtp_get_capabilities_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_set_configuration_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_get_configuration_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_re_configure_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_open_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_start_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_close_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_suspend_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_abort_handler(struct bt_avdtp *session,
+			struct net_buf *buf, uint8_t msg_type, uint8_t tid);
+static void avdtp_timeout(struct k_work *work);
+static struct net_buf *avdtp_create_pdu(uint8_t msg_type,
+					uint8_t pkt_type,
+					uint8_t sig_id);
+
+enum seid_state {
+	AVDTP_IDLE = 0,
+	AVDTP_CONFIGURED,
+	/* establishing the transport sessions. */
+	AVDTP_OPENING,
+	AVDTP_OPEN,
+	AVDTP_STREAMING,
+	AVDTP_CLOSING,
+	AVDTP_ABORTING,
+};
+
 static const struct {
 	uint8_t sig_id;
 	void (*func)(struct bt_avdtp *session, struct net_buf *buf,
-		     uint8_t msg_type);
+			 uint8_t msg_type, uint8_t tid);
 } handler[] = {
+	{BT_AVDTP_DISCOVER, avdtp_discover_handler},
+	{BT_AVDTP_GET_CAPABILITIES, avdtp_get_capabilities_handler},
+	{BT_AVDTP_SET_CONFIGURATION, avdtp_set_configuration_handler},
+	{BT_AVDTP_GET_CONFIGURATION, avdtp_get_configuration_handler},
+	{BT_AVDTP_RECONFIGURE, avdtp_re_configure_handler},
+	{BT_AVDTP_OPEN, avdtp_open_handler},
+	{BT_AVDTP_START, avdtp_start_handler},
+	{BT_AVDTP_CLOSE, avdtp_close_handler},
+	{BT_AVDTP_SUSPEND, avdtp_suspend_handler},
+	{BT_AVDTP_ABORT, avdtp_abort_handler},
 };
 
+
+/* L2CAP Interface callbacks */
+void bt_avdtp_media_l2cap_connected(struct bt_l2cap_chan *chan)
+{
+	struct bt_avdtp *session;
+	struct bt_avdtp_seid_lsep *lsep =
+		CONTAINER_OF(chan, struct bt_avdtp_seid_lsep, media_br_chan.chan);
+
+	session = lsep->session;
+	if (session == NULL) {
+		return;
+	}
+
+	LOG_DBG("chan %p session %p", chan, session);
+	lsep->seid_state = AVDTP_OPEN;
+	if ((session->req != NULL) && (session->req->func != NULL)) {
+		struct bt_avdtp_req *req = session->req;
+
+		if (!chan) {
+			LOG_ERR("Invalid AVDTP chan");
```

Comment:
if the condition (!chan) is ture, it means "lsep" is invalid. The logic here is contradictory.

Reply:
Fixed, Thanks @lylezhu2012